floppyd_installtest.1
'\" t
.TH floppyd_installtest 1 "02Dec18" mtools-4.0.22
.SH Name
floppyd_installtest - tests whether floppyd is installed and running
'\" t
.de TQ
.br
.ns
.TP \\$1
..

.tr \(is'
.tr \(if`
.tr \(pd"

.SH Note\ of\ warning
This manpage has been automatically generated from mtools's texinfo
documentation, and may not be entirely accurate or complete.  See the
end of this man page for details.
.PP
.SH Description
.PP
\&\fR\&\f(CWFloppyd_installtest\fR is used to check for the presence of a running
floppyd daemon. This is useful, if you have a small front-end script to
mtools, which decides whether to use floppyd or not.
.PP
\&\fR\&\f(CWfloppyd_installtest\fR [\fR\&\f(CW-f\fR]  Connect-String
.PP
If the \fR\&\f(CW-f\fR option is specified, \fR\&\f(CWfloppyd_installtest\fR does a
full X-Cookie authentication and complains if this does not work.
.PP
The connect-String has the format described in the floppyd-section:
\&\fIhostname\fR\fR\&\f(CW:\fR\fIdisplaynumber\fR[\fR\&\f(CW/\fR\fIbaseport\fR]
.PP
.SH See\ Also
Mtools' texinfo doc
.SH Viewing\ the\ texi\ doc
This manpage has been automatically generated from mtools's texinfo
documentation. However, this process is only approximative, and some
items, such as crossreferences, footnotes and indices are lost in this
translation process.  Indeed, these items have no appropriate
representation in the manpage format.  Moreover, not all information has
been translated into the manpage version.  Thus I strongly advise you to
use the original texinfo doc.  See the end of this manpage for
instructions how to view the texinfo doc.
.TP
* \ \ 
To generate a printable copy from the texinfo doc, run the following
commands:
 
.nf
.ft 3
.in +0.3i
    ./configure; make dvi; dvips mtools.dvi
.fi
.in -0.3i
.ft R
.PP
 
\&\fR
.TP
* \ \ 
To generate a html copy,  run:
 
.nf
.ft 3
.in +0.3i
    ./configure; make html
.fi
.in -0.3i
.ft R
.PP
 
\&\fRA premade html can be found at
\&\fR\&\f(CW\(ifhttp://www.gnu.org/software/mtools/manual/mtools.html\(is\fR
.TP
* \ \ 
To generate an info copy (browsable using emacs' info mode), run:
 
.nf
.ft 3
.in +0.3i
    ./configure; make info
.fi
.in -0.3i
.ft R
.PP
 
\&\fR
.PP
The texinfo doc looks most pretty when printed or as html.  Indeed, in
the info version certain examples are difficult to read due to the
quoting conventions used in info.
.PP


NEWS
v4_0_22
	- Fixed -f flag for mformat (size is KBytes, rather than sectors)
	- Fixed toupper/tolower usage (unsigned char rather than plain signed)
v4_0_21
	- Fixed compilation for MingW
	- After MingW compilation, make sure executable has .exe extension
	- Addressed compiler warnings
	- Fixed length handling in character set conversion (Unicode file names)
	- Fixed matching of character range, when containing Unicode
	characters (mdir "c:test[α-ω].exe")
	- Fixed initialization of my_scsi_cmd constructor
v4_0_20
	- initialize directory entries to 0
	- bad message "Too few sectors" replaced with "Too many sectors"
	- apostrophe in mlabel no longer causes generation of long entry
	- option to fake system date for file creation using the
	SOURCE_DATE_EPOCH environment variables

	- can now be compiled with "clang" compiler
	- fallback function for strndup, for those platforms that don't have it
	- fixed a number of -Wextra warnings

	- new compressed archive formats for uz/lz

	- allow to specify number of reserved sectors for FAT32.
	- file/device locking with timeout (rather than immediate failure)
	- fixed support for BPB-less legacy formats.
	- removed check that disk must be an integer number of tracks.

	- removed .eh/.oh macros from manual pages
v4_0_19
	- Fix for short file names starting with character 0xE5
	(by remapping it to 0x5)
	- mpartition: Partition types closer to what Microsoft uses
	- mformat: figure out LBA geometry as last resort if geometry
	is neither specified in config and/or commandline, nor can be
	queried from the device
	- mformat: use same default cluster size by size as Microsoft
	for FAT32
	- additional sanity checks
	- document how cluster size is picked in mformat.c man page
	- document how partition types are picked in mpartition.c man page
v4_0_18	
	Fix for names of iconv encodings on AIX
	Fix mt_size_t on NetBSD
	Fixed compilation on Mingw
	Fixed doc (especially mformat)
	Fix mformat'ing of FAT12 filesystems with huge cluster sizes
	Minfo prints image file name in mformat command line if an image
	file name was given
	Always generate gzip-compressed RPMs, in order to remain
	compatible with older distributions
	Fixed buffer overflow with drive letter in mclasserase
v4_0_17
	mbadblocks now takes a list of bad blocks (either as sectors
	or as clusters)
	mbadblocks now is able to do write scanning for bad blocks
	mshowfat can show cluster of specific offset
	Enable mtools to deal with very small sector sizes...
	Fixed encoding of all-lowercase names (no need to mangle
	these)
	Consider every directory entry after an ENDMARK (0x00) to be deleted
	After writing a new entry at end of a directory, be sure to also add
	an ENDMARK (0x00)

	Deal with possibility of a NULL pointer being returned by
	localtime during timestamp conversion
v4_0_16
	configure.in fixes
	fixed formatting of fat_size_calculation.tex document
	compatibility with current autoconfig versions
	Make it clear that label is limited to 11 characters
	Fixed typo in initialization of FAT32 info sector
v4_0_15
	Added missing -i option to mshortname
	Splitted .deb package into mtools and floppyd in order to
	match Ubuntu
v4_0_14
	New mshortname command
	Fix floppyd for disks bigger than 2 Gig
	Remove obsolete -z flag
	Remove now unsupported AC_USE_SYSTEM_EXTENSIONS
	Fixed output formatting of mdir if MTOOLS_DOTTED_DIR is set
	Mformat now correctly writes backup boot sector
	Fixed signedness of serial number in mlabel
	Fixed buffer size problem in mlabel
	Make mlabel write backup boot sector if FAT32
	Catch situation where both clear and new label are given to mlabel
	Quote filename parameters to scripts
	Mformat: Close file descriptor for boot sector
	Added lzip support to scripts/uz
	Added Tot_sectors option to mformat
	Fixed hidden sector handling in mformat
	Minfo generates mformat command lines containing new -T option
	Mlabel prints error if label too long
v4_0_13
	Merged Debian patches
v4_0_12
	Mingw compatibility fixes
v4_0_11
	Fixed compiler warnings in mlabel.c and elsewhere
	Fixed h flag in mattrib.c
	Added missing error checking in floppyd and elsewhere
	
v4_0_10
	More copyright stuff...
	Fixed issues with max filesize (was 2GB instead of 4GB, and
	warned only after copying the beginning)
v4_0_9
	More copyright stuff
v4_0_8
	Corrected copyright attributions in the various files
v4_0_7
	Fixed conversion to native on OS/2
	Fix parsing of --help flag
v4_0_6
	Fallback for missing wchar_t iconv codepage on OS/2
	Fixes for LSEEK64 support
	Support for --help that returns a 0 exit status
v4_0_5
	Make setpgrp() usage in floppyd conditional
	Re-instate PACKED around structure (ARM)
	LSEEK64
	
v4_0_4
	BSD support: SCSI, use getuserid/getgroupid in floppyd
	Another attempt at putwc fix for OS/2
	Further GNU fixes
	Fallback for putwc if there is wchar (OS/2)
v4_0_3
	Fix multipart pathname parsing bug in vfat.c (forgot limited length)
	Supplied fallback define for putwc
	Copyright notices in all sources
v4_0_2
	Off-by-2 error in unix_name in file_name.c
v4_0_1
	Missing functions on Solaris
v4_0_0
	Offset for -i-specified image files
v4_0_0_pre2
	Use transliteration to represent characters which don't exist in
	target set
v4_0_0_pre1
	Mtools-4 with Unicode support
	Released 4.0.0_pre1
v20071226
	Debian build files
	Fixed security issue with doctored file names
	64 bit compilation fixes
v20070601
	Fixed misc blunders...
v20070531
	Fixed lots of minor items raised by gcc4
	Merged some of the BSD patches
	New version of amuFormat.sh written in sh rather than csh
	Support for config parameters after -i file
	Released 3.9.11
v20070411
	Added sizecode printing on minfo
	In mformat manpage, use same flag for sector number than printed in
	minfo
	Limit sizecode to 6, else it will overflow max sector size defined in
	msdos.h
v20070308
	Applied mingw patch by Jamey Sharp and Josh Triplett
v20070306
	Fixed doc about /etc/default
v20070305
	Fixed mlabel on read-only disks
v20060626
	Merged Redhat/Fedora patches

floppyd_installtest.c
/*  Copyright 1999 Peter Schlaile.
 *  Copyright 1999-2002,2006,2007,2009 Alain Knaff.
 *  This file is part of mtools.
 *
 *  Mtools is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  Mtools is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with Mtools.  If not, see <http://www.gnu.org/licenses/>.
 *
 * Small install-test utility to check if a floppyd-server is running on the
 * X-Server-Host.
 *
 * written by:
 *
 * Peter Schlaile
 *
 * udbz@rz.uni-karlsruhe.de
 *
 */

#include "sysincludes.h"
#include "stream.h"
#include "mtools.h"
#include "msdos.h"
#include "scsi.h"
#include "partition.h"
#include "floppyd_io.h"

#ifdef USE_FLOPPYD
#include <sys/socket.h>
#include <arpa/inet.h>
#include <netdb.h>

/* ######################################################################## */

typedef unsigned char Byte;
typedef unsigned long Dword;
typedef mt_off_t Qword;

static const char* AuthErrors[] = {
	"Auth success!",
	"Auth failed: Packet oversized!",
	"Auth failed: X-Cookie doesn't match!",
	"Auth failed: Wrong transmission protocol version!",
	"Auth failed: Device locked!"
};

#include "byte_dword.h"
#include "read_dword.h"

static int write_dword(int handle, Dword parm)
{
	Byte val[4];

	dword2byte(parm, val);

	if(write(handle, val, 4) < 4)
		return -1;
	return 0;
}


/* ######################################################################## */

static int authenticate_to_floppyd(char fullauth, int sock, char *display, 
				   int protoversion)
{
	off_t filelen=0;
	Byte buf[16];
	const char *command[] = { "xauth", "xauth", "extract", "-", 0, 0 };
	char *xcookie = NULL;
	Dword errcode;
	int bytesRead;
	int cap=0;

	if (fullauth) {
		command[4] = display;

		filelen=strlen(display);
		filelen += 100;

		xcookie = (char *) safe_malloc(filelen+4);
		filelen = safePopenOut(command, xcookie+4, filelen);
		if(filelen < 1)
		    return AUTH_AUTHFAILED;
	}
	dword2byte(4,buf);
	dword2byte(protoversion,buf+4);
	if(write(sock, buf, 8) < 8)
		return AUTH_IO_ERROR;

	bytesRead = read_dword(sock);

	if (bytesRead != 4 && bytesRead != 12) {
		return AUTH_WRONGVERSION;
	}


	errcode = read_dword(sock);

	if (errcode != AUTH_SUCCESS) {
		return errcode;
	}

	protoversion = FLOPPYD_PROTOCOL_VERSION_OLD;	
	if(bytesRead >= 12) {
	    protoversion = read_dword(sock);
	    cap = read_dword(sock);
	}
	
	fprintf(stderr, "Protocol Version=%d\n", protoversion);
	if(protoversion >= FLOPPYD_PROTOCOL_VERSION) {
	  fprintf(stderr, "Capabilities:%s%s\n",		  
		  (cap & FLOPPYD_CAP_EXPLICIT_OPEN) ? " ExplicitOpen" : "",
		  (cap & FLOPPYD_CAP_LARGE_SEEK) ? " LargeFiles" : "");
	}

	if (fullauth) {
		dword2byte(filelen, (Byte *) xcookie);
		if(write(sock, xcookie, filelen+4) < filelen+4)
			return AUTH_IO_ERROR;

		if (read_dword(sock) != 4) {
			return AUTH_PACKETOVERSIZE;
		}

		errcode = read_dword(sock);
	}

	return errcode;

}


/* ######################################################################## */

static int get_host_and_port(const char* name, char** hostname, char **display,
			     short* port)
{
	char* newname = strdup(name);
	char* p;
	char* p2;

	p = newname;
	while (*p != '/' && *p) p++;
	p2 = p;
	if (*p) p++;
	*p2 = 0;
	
	*port = atoi(p);
	if (*port == 0) {
		*port = FLOPPYD_DEFAULT_PORT;	
	}

	*display = strdup(newname);

	p = newname;
	while (*p != ':' && *p) p++;
	p2 = p;
	if (*p) p++;
	*p2 = 0;

	*port += atoi(p);  /* add display number to the port */

	if (!*newname || strcmp(newname, "unix") == 0) {
		free(newname);
		newname = strdup("localhost");
	}

	*hostname = newname;
	return 1;
}

/*
 *  * Return the IP address of the specified host.
 *  */
static IPaddr_t getipaddress(char *ipaddr)
{
	
	struct hostent  *host;
	IPaddr_t        ip;
	
	if (((ip = inet_addr(ipaddr)) == INADDR_NONE) &&
	    (strcmp(ipaddr, "255.255.255.255") != 0)) {
		
		if ((host = gethostbyname(ipaddr)) != NULL) {
			memcpy(&ip, host->h_addr, sizeof(ip));
		}
		
		endhostent();
	}
	
#ifdef DEBUG
	fprintf(stderr, "IP lookup %s -> 0x%08lx\n", ipaddr, ip);
#endif
	  
	return (ip);
}

/*
 *  * Connect to the floppyd server.
 *  */
static int connect_to_server(IPaddr_t ip, short port)
{
	
	struct sockaddr_in      addr;
	int                     sock;
	
	/*
	 * Allocate a socket.
	 */
	if ((sock = socket(AF_INET, SOCK_STREAM, 0)) < 0) {
		return (-1);
	}
	
	/*
	 * Set the address to connect to.
	 */
	
	addr.sin_family = AF_INET;
	addr.sin_port = htons(port);
	addr.sin_addr.s_addr = ip;
	
        /*
	 * Connect our socket to the above address.
	 */
	if (connect(sock, (struct sockaddr *)&addr, sizeof(addr)) < 0) {
		return (-1);
	}

        /*
	 * Set the keepalive socket option to on.
	 */
	{
		int             on = 1;
		setsockopt(STDIN_FILENO, SOL_SOCKET, SO_KEEPALIVE, 
			   (char *)&on, sizeof(on));

	}
	
	return (sock);
}

int main (int argc, char** argv) 
{
	char* hostname;
	char* display;
	char* name;
	short port;
	int sock;
	int reply;
	int rval;
	int protoversion;
	char fullauth = 0;
	Byte opcode = OP_CLOSE;

        if (argc < 2) {
	       puts("Usage: floppyd_installtest [-f] Connect-String\n"
		    "-f\tDo full X-Cookie-Authentication");
	       return -1;
	}

	name = argv[1];
	if (strcmp(name, "-f") == 0) {
		fullauth = 1;
		name = argv[2];
	}

	rval = get_host_and_port(name, &hostname, &display, &port);
	
	if (!rval) return -1;

	sock = connect_to_server(getipaddress(hostname), port);

	if (sock == -1) {
		fprintf(stderr,
			"Can't connect to floppyd server on %s, port %i!\n",
			hostname, port);
		return -1;
	}
	
	protoversion = FLOPPYD_PROTOCOL_VERSION;
	while(1) {
	    reply = authenticate_to_floppyd(fullauth, sock, display,
					    protoversion);
	    if(protoversion == FLOPPYD_PROTOCOL_VERSION_OLD)
		break;
	    if(reply == AUTH_WRONGVERSION) {
		/* fall back on old version */
		protoversion = FLOPPYD_PROTOCOL_VERSION_OLD;
		continue;
	    }
	    break;
	}

	if (reply != 0) {
		fprintf(stderr, 
			"Connection to floppyd failed:\n"
			"%s\n", AuthErrors[reply]);
		return -1;
	}
	
	free(hostname);
	free(display);

	if(write_dword(sock, 1) < 0) {
		fprintf(stderr,
			"Short write to floppyd:\n"
			"%s\n", strerror(errno));
	}

	if(write(sock, &opcode, 1) < 0) {
		fprintf(stderr,
			"Short write to floppyd:\n"
			"%s\n", strerror(errno));
	}

	close(sock);

	return 0;
}
#endif
